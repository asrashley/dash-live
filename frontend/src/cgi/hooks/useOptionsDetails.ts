import { type ReadonlySignal, useSignal } from "@preact/signals";

import { fieldGroups } from "@dashlive/options";

import { InputOptionName } from "../types/InputOptionName";

function createOptionNames(): InputOptionName[] {
    const names: InputOptionName[] = [];
    fieldGroups.forEach(grp => {
        grp.fields.forEach(({ name, fullName, shortName }) => {
            names.push({
                cgiName: name,
                fullName,
                shortName,
            });
        });
    });
    return names;
}

export interface UseOptionsDetailsHook {
    allOptions: ReadonlySignal<InputOptionName[]>;
}

export function useOptionsDetails(): UseOptionsDetailsHook {
    const allOptions = useSignal<InputOptionName[]>(createOptionNames());
    return { allOptions };
}