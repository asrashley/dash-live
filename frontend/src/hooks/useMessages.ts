import { type ReadonlySignal, signal, useComputed } from "@preact/signals";
import { MessageLevel, MessageType } from "../types/MessageType";

type MessagesState = {
  alerts: MessageType[];
  nextId: number;
}

const messages = signal<MessagesState>({ alerts: [], nextId: 1 });

// only exported for use in tests
export function resetAllMessages() {
  messages.value = { alerts: [], nextId: 1 };
}

export type AppendMessageFn = (level: MessageLevel, text: string) => number;

const appendMessage: AppendMessageFn = (level: MessageLevel, text: string) => {
  const id = messages.value.nextId;
  const alerts = [
    ...messages.value.alerts,
    {
      text,
      id,
      level,
    },
  ];
  messages.value = {
    nextId: id + 1,
    alerts,
  };
  return id;
};

const removeAlert = (id: number) => {
  const alerts = messages.value.alerts.filter((a) => a.id !== id);
  messages.value = { ...messages.value, alerts };
};

export interface UseMessagesHook {
  alerts: ReadonlySignal<MessageType[]>
  appendMessage: AppendMessageFn;
  removeAlert: (id: number) => void;
}

export function useMessages(): UseMessagesHook {
  const alerts = useComputed(() => {
    return messages.value.alerts;
  });

  return { alerts, appendMessage, removeAlert };
}
