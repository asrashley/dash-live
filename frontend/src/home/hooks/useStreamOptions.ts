import { useCallback } from "preact/hooks";
import { type ReadonlySignal, type Signal, useComputed, useSignal } from "@preact/signals";

import { defaultCgiOptions, drmSystems } from "@dashlive/options";
import { CombinedStream } from "../../hooks/useCombinedStreams";
import { InputFormData } from "../../types/InputFormData";

const keyName = "dashlive.homepage.options";

function getDefaultOptions(): InputFormData {
  const lsKey = localStorage.getItem(keyName);
  const previousOptions = lsKey ? JSON.parse(lsKey) : {};
  return {
    ...defaultCgiOptions,
    manifest: "hand_made.mpd",
    mode: "vod",
    stream: undefined,
    ...previousOptions,
  };
}

const skipKeys = new RegExp(`^(${[
  ...drmSystems,
  ...drmSystems.map(name => `${name}__enabled`),
  ...drmSystems.map(name => `${name}__drmloc`),
  'drms', 'manifest', 'stream', 'mode'].join('|')})$`);

const manifestSkipKeys = /^(player|dashjs|shaka)/;

export type EnabledDrmSystems = {
  [name: string]: boolean;
};

export interface UseStreamOptionsHook {
  data: Signal<InputFormData>;
  drms: ReadonlySignal<EnabledDrmSystems>;
  stream: ReadonlySignal<CombinedStream>;
  mode: ReadonlySignal<string>;
  manifest: ReadonlySignal<string>;
  nonDefaultOptions: ReadonlySignal<object>;
  manifestOptions: ReadonlySignal<object>;
  setValue: (name: string, value: string | number | boolean) => void;
  resetAllValues: () => void;
}

const emptyStream: CombinedStream = {
  title: "",
  value: "",
  mps: false
}

export interface UseStreamOptionsProps {
  streamNames: ReadonlySignal<string[]>;
  streamsMap: ReadonlySignal<Map<string, CombinedStream>>;
}
export function useStreamOptions({ streamNames, streamsMap }: UseStreamOptionsProps): UseStreamOptionsHook {
  const data = useSignal<InputFormData>(getDefaultOptions());
  const stream = useComputed<CombinedStream>(() => {
    const name: string = (data.value['stream'] as string | undefined) ?? streamNames.value[0];
    return streamsMap.value.get(name) ?? emptyStream;
  });
  const mode = useComputed<string>(() => data.value['mode'] as string);
  const manifest = useComputed<string>(() => data.value['manifest'] as string);
  const drms = useComputed<EnabledDrmSystems>(() => Object.fromEntries(drmSystems.map(name => [name, data.value[name] === "1"])));
  const nonDefaultOptions = useComputed<object>(() => {
    const params = Object.entries(data.value)
      .filter(([key, value]) => defaultCgiOptions[key] != value)
      .filter(([key]) => !skipKeys.test(key));
    const drm = drmSystems.filter(system => data.value[system] === "1").map(system => {
      const drmLoc = data.value[`${system}__drmloc`];
      return drmLoc ? `${system}-${drmLoc}`: system;
    });
    if (drm.length) {
      params.push(['drm', drm.join(',')]);
    }
    return Object.fromEntries(params);
  });
  const manifestOptions = useComputed<object>(() => Object.fromEntries(
    Object.entries(nonDefaultOptions.value).filter(([key]) => !manifestSkipKeys.test(key))));

  const setValue = useCallback(
    (name: string, value: string | number | boolean) => {
      if (value === true) {
        value = "1";
      } else if (value === false) {
        value = "0";
      }
      data.value = {
        ...data.value,
        [name]: value,
      };
      if (name === "stream" && mode.value === "odvod") {
        const nextStream = streamsMap.value.get(value as string);
        if (nextStream?.mps) {
          data.value = {
            ...data.value,
            mode: "vod",
          };
        }
      }
      const params = Object.fromEntries(
        Object.entries(data.value).filter(
          ([key, value]) => defaultCgiOptions[key] !== value
        )
      );
      localStorage.setItem(keyName, JSON.stringify(params));
    },
    [data, mode, streamsMap]
  );

  const resetAllValues = useCallback(() => {
    data.value = { ...getDefaultOptions() };
    localStorage.removeItem(keyName);
  }, [data]);

  return { data, drms, stream, mode, manifest, nonDefaultOptions, manifestOptions, setValue, resetAllValues };
}
