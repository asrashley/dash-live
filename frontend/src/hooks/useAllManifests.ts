import { createContext } from "preact";
import { useContext, useEffect } from "preact/hooks";
import { Signal, useComputed, useSignal } from "@preact/signals";

import { EndpointContext } from "../endpoints";
import { AllManifests } from "../types/AllManifests";


export interface UseAllManifestsHooks {
  allManifests: Signal<AllManifests | undefined>;
  names: Signal<string[]>;
  error: Signal<string | null>;
}

export const AllManifestsContext = createContext<UseAllManifestsHooks>(null);

export function useAllManifests(): UseAllManifestsHooks {
  const apiRequests = useContext(EndpointContext);
  const allManifests = useSignal<AllManifests | undefined>();
  const error = useSignal<string | null>(null);
  const names = useComputed<string[]>(() => {
    const keys = allManifests.value ? [...Object.keys(allManifests.value)]: [];
    keys.sort();
    return keys;
  });

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const fetchAllManifestsIfRequired = async () => {
      if (!allManifests.value) {
        try {
          const data = await apiRequests.getAllManifests({
            signal,
          });
          if (!signal.aborted) {
            allManifests.value = data;
            error.value = null;
          }
        } catch (err) {
          if (!signal.aborted) {
            console.error(err);
            error.value = `${err}`;
          }
        }
      }
    };
    fetchAllManifestsIfRequired();

    return () => {
      controller.abort();
    };
  }, [apiRequests, allManifests, error]);

  return { allManifests, names, error };
}
