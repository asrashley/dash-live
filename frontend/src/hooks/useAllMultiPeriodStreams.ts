import { createContext } from "preact";
import { useEffect, useState, useContext, useCallback } from "preact/hooks";
import { type ReadonlySignal, useSignal } from "@preact/signals";

import { EndpointContext } from "../endpoints";
import { AllMultiPeriodStreamsJson, MultiPeriodStreamSummary } from "../types/AllMultiPeriodStreams";

export interface UseAllMultiPeriodStreamsHook {
  error: ReadonlySignal<string | null>;
  streams: ReadonlySignal<MultiPeriodStreamSummary[]>;
  loaded: ReadonlySignal<boolean>;
  sort: (field: string, ascending: boolean) => void;
  sortField: string;
  sortAscending: boolean;
}

export const AllMultiPeriodStreamsContext = createContext<UseAllMultiPeriodStreamsHook>(null);

function sortStreams(streams: ReadonlySignal<MultiPeriodStreamSummary[]>, field: string, ascending: boolean): MultiPeriodStreamSummary[] {
  const newOrder = [...streams.value];
  newOrder.sort((a, b) => {
    const left = a[field];
    const right = b[field];
    if (left === right) {
      return 0;
    }
    if (left < right) {
      return ascending ? -1 : 1;
    }
    return ascending ? 1 : -1;
  });
  return newOrder;
}

export function useAllMultiPeriodStreams(): UseAllMultiPeriodStreamsHook {
  const apiRequests = useContext(EndpointContext);
  const streams = useSignal<MultiPeriodStreamSummary[]>([]);
  const loaded = useSignal<boolean>(false);
  const error = useSignal<string | null>(null);
  const [sortField, setSortField] = useState<string>("name");
  const [sortAscending, setSortAscending] = useState<boolean>(true);

  const sort = useCallback(
    (field: string, ascending: boolean) => {
      setSortField(field);
      setSortAscending(ascending);
      streams.value = sortStreams(streams, field, ascending);
    },
    [streams]
  );

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const fetchStreamsIfRequired = async () => {
      if (!loaded.value) {
        try {
          const data: AllMultiPeriodStreamsJson = await apiRequests.getAllMultiPeriodStreams({ signal });
          if (!signal.aborted) {
            loaded.value = true;
            streams.value = data.streams;
            error.value = null;
            sortStreams(streams, sortField, sortAscending);
          }
        } catch (err) {
          if (!signal.aborted) {
            error.value = `Fetching multi-period streams list failed: ${err}`;
          }
        }
      }
    };

    fetchStreamsIfRequired();

    return () => {
      controller.abort();
    };
  }, [apiRequests, error, loaded, sortAscending, sortField, streams]);

  return { error, streams, loaded, sort, sortField, sortAscending };
}
