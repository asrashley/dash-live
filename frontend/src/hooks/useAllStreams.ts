import { createContext } from "preact";
import { useContext, useEffect } from "preact/hooks";
import { useSignal, useComputed, type Signal } from "@preact/signals";

import { EndpointContext } from "../endpoints";
import { Stream } from "../types/Stream";
import { StreamTrack } from "../types/StreamTrack";
import { DecoratedStream } from "../types/DecoratedStream";

function findStreamTracks(stream: Stream): StreamTrack[] {
  const tracks = new Map();
  for (const mf of stream.media_files) {
    const { track_id, content_type, codec_fourcc } = mf;
    let track: StreamTrack = tracks.get(track_id);
    if (track === undefined) {
      track = {
        clearBitrates: 0,
        encryptedBitrates: 0,
        track_id,
        content_type,
        codec_fourcc,
      };
      tracks.set(track_id, track);
    }
    if (mf.encrypted) {
      track.encryptedBitrates += 1;
    } else {
      track.clearBitrates += 1;
    }
  }
  const ids = [...tracks.keys()];
  ids.sort();
  return ids.map((tid) => tracks.get(tid));
}

export function decorateAllStreams(streams: Stream[]): DecoratedStream[] {
  return streams.map((stream) => {
    const decoratedStream: DecoratedStream = {
      ...stream,
      tracks: findStreamTracks(stream),
    };
    return decoratedStream;
  });
}

export interface UseAllStreamsHook {
  allStreams: Signal<DecoratedStream[]>,
  loaded: Signal<boolean>,
  streamsMap: Signal<Map<string, DecoratedStream>>,
  error: Signal<string | null>;
}

export const AllStreamsContext = createContext<UseAllStreamsHook>(null);

export function useAllStreams(): UseAllStreamsHook {
  const apiRequests = useContext(EndpointContext);
  const streams = useSignal<DecoratedStream[] | undefined>();
  const loaded = useSignal<boolean>(false);
  const error = useSignal<string | null>(null);
  const allStreams = useComputed<DecoratedStream[]>(() => streams.value ?? []);
  const streamsMap = useComputed<Map<string, DecoratedStream>>(() => {
    const rv = new Map();
    if (allStreams.value) {
      for (const stream of allStreams.value) {
        rv.set(stream.pk, stream);
      }
    }
    return rv;
  });

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const fetchAllStreamsIfRequired = async () => {
      if (!streams.value) {
        try {
          const data = await apiRequests.getAllStreams({
            signal,
            withDetails: true,
          });
          if (!signal.aborted) {
            streams.value = decorateAllStreams(data.streams);
            error.value = null;
            loaded.value = true;
          }
        } catch (err) {
          if (!signal.aborted) {
            console.error(err);
            error.value = `${err}`;
          }
        }
      }
    };

    fetchAllStreamsIfRequired();

    return () => {
      controller.abort();
    };
  }, [apiRequests, error, loaded, streams]);

  return { allStreams, loaded, streamsMap, error };
}
