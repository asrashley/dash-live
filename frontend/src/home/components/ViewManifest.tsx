import {
  type ReadonlySignal,
  useSignal,
  useSignalEffect,
} from "@preact/signals";
import { useState } from "preact/hooks";

import { useMessages } from "../../hooks/useMessages";
import { LoadingSpinner } from "../../components/LoadingSpinner";

export interface ViewManifestProps {
    manifestUrl: ReadonlySignal<URL>;
}

export function ViewManifest({ manifestUrl }: ViewManifestProps) {
  const loadedUrl = useSignal<string>("");
  const [xmlText, setXmlText] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(true);
  const { appendMessage } = useMessages();

  useSignalEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const fetchManifestIfRequired = async (url: string) => {
      if (url && loadedUrl.value !== url) {
        const headers = new Headers({
          accept: "application/dash+xml",
        });
        try {
          setLoading(true);
          const response = await fetch(url, {
            headers,
            signal,
          });
          loadedUrl.value = url;
          if (response.ok) {
            const parser = new DOMParser();
            const doc = parser.parseFromString(
              await response.text(),
              "text/xml"
            );
            const ser = new XMLSerializer();
            setXmlText(ser.serializeToString(doc));
          } else {
            appendMessage(
              "danger",
              `Fetching manifest failed: ${response.status}: ${response.statusText}`
            );
            setXmlText(`Fetching manifest failed: ${response.status}: ${response.statusText}`);
          }
        } catch (err) {
          if (!signal.aborted) {
            appendMessage("danger", `Fetching manifest failed: ${err}`);
            setXmlText(`Fetching manifest failed: ${err}`);
          }
        } finally {
            setLoading(false);
        }
      }
    };
    const fetchUrl = manifestUrl.value.href;
    fetchManifestIfRequired(fetchUrl);
    return () => {
      if (fetchUrl !== manifestUrl.value.href) {
        controller.abort("URL changed");
      }
    };
  });

  return (
    <div className="display-manifest">
      {loading ? <LoadingSpinner /> : <pre id="manifest-xml">{xmlText}</pre>}
    </div>
  );
}
