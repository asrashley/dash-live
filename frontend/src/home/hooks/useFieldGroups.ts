import { useContext } from "preact/hooks";
import { type ReadonlySignal, useComputed } from "@preact/signals";

import { fieldGroups, drmSystems } from '@dashlive/options';

import { useAllManifests  } from '../../hooks/useAllManifests';
import { UseCombinedStreams  } from '../../hooks/useCombinedStreams';
import { StaticInputProps } from "../../form/types/StaticInputProps";
import { InputFormGroup } from "../../form/types/InputFormGroup";

const drmSkipNames = new RegExp(`^${drmSystems.map(name => `${name}__enabled`).join('|')}$`);

interface UseFieldGroupsHook {
  homeFieldGroups: ReadonlySignal<InputFormGroup[]>;
}

export function useFieldGroups(): UseFieldGroupsHook {
  const { allManifests, names } = useAllManifests();
  const { streamNames, streamsMap } = useContext(UseCombinedStreams);

  const homeFieldGroups = useComputed<InputFormGroup[]>(() => {
    const playbackMode: StaticInputProps = {
      name: "mode",
      shortName: "mode",
      fullName: "playbackMode",
      title: "Playback Mode",
      type: "radio",
      options: [
        {
          title: "Video On Demand (using live profile)",
          value: "vod",
        },
        {
          title: "Live stream (using live profile)",
          value: "live",
        },
        {
          title: "Video On Demand (using on-demand profile)",
          value: "odvod",
        },
      ],
    };
    const selectManifest: StaticInputProps = {
      name: "manifest",
      shortName: "manifest",
      fullName: "manifest",
      title: "Manifest",
      text: "Manifest template to use",
      type: "select",
      options: names.value.map((name: string) => {
        const msft = allManifests.value[name];
        return {
          title: msft.title,
          value: name,
        };
      }),
    };
    const selectStream: StaticInputProps = {
      name: "stream",
      shortName: "stream",
      fullName: "stream",
      title: "Stream",
      text: "Stream to play",
      type: "select",
      options: streamNames.value.map((value: string) => ({
        title: streamsMap.value.get(value).title,
        value,
      })),
    };
    const selectDrmSystem: StaticInputProps = {
      name: "drms",
      shortName: "drms",
      fullName: "drms",
      title: "DRM systems",
      text: "DRM systems to enable for encrypted AdaptationSets",
      type: "multiselect",
      options: drmSystems.map((name: string) => ({
        name,
        title: name,
        value: name,
      })),
    };
    const generalOptions: InputFormGroup = {
      ...fieldGroups[0],
      fields: [
        playbackMode,
        selectManifest,
        selectStream,
        selectDrmSystem,
        ...fieldGroups[0].fields,
      ],
    };
    const otherGroups: InputFormGroup[] = fieldGroups.slice(1).map(grp => ({
        ...grp,
        fields: grp.fields.filter(({name}) => !drmSkipNames.test(name)),
    }));
    return [generalOptions, ...otherGroups];
  });

  return { homeFieldGroups };
}
