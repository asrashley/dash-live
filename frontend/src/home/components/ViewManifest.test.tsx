import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { mock } from "vitest-mock-extended";
import fetchMock from "@fetch-mock/vitest";
import { signal } from "@preact/signals";
import log from "loglevel";

import { renderWithProviders } from "../../test/renderWithProviders";
import { FakeEndpoint } from "../../test/FakeEndpoint";
import { MockDashServer } from "../../test/MockServer";
import { useMessages, UseMessagesHook } from "../../hooks/useMessages";

import { ViewManifest } from "./ViewManifest";

vi.mock("../../hooks/useMessages", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useMessages: vi.fn(),
  };
});

describe("ViewManifest component", () => {
  const useMessagesMock = vi.mocked(useMessages);
  const messagesMock = mock<UseMessagesHook>();
  const initialUrl = signal<URL>();
  let endpoint: FakeEndpoint;
  //let server: MockDashServer;

  beforeEach(() => {
    useMessagesMock.mockReturnValue(messagesMock);
    endpoint = new FakeEndpoint(document.location.origin);
    new MockDashServer({
      endpoint,
    });
    initialUrl.value = new URL(
      "/dash/vod/bbb/hand_made.mpd",
      document.location.href
    );
  });

  afterEach(() => {
    endpoint.shutdown();
    vi.clearAllMocks();
    fetchMock.mockReset();
  });

  test("matches snapshot", async () => {
    const { asFragment, findByText, getBySelector } = renderWithProviders(
      <ViewManifest url={initialUrl} />
    );
    const inp = getBySelector("#id_mpd_url") as HTMLInputElement;
    expect(inp.value).toEqual(initialUrl.value.href);
    await findByText("urn:mpeg:dash:profile:isoff-live:2011", { exact: false });
    expect(asFragment).toMatchSnapshot();
  });

  test("shows error if manifest fetch fails", async () => {
    endpoint.setServerStatus(500);
    const prom = endpoint.addResponsePromise("get", initialUrl.value.pathname);
    const { findByText } = renderWithProviders(
      <ViewManifest url={initialUrl} />
    );
    await expect(prom).resolves.toEqual(500);
    await findByText("Fetching manifest failed", { exact: false });
    expect(messagesMock.appendMessage).toHaveBeenCalledTimes(1);
  });
});
