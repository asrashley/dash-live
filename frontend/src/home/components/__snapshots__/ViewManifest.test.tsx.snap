// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ViewManifest component > matches snapshot 1`] = `[Function]`;
